File: src/lib/objmdl/mdl/components/node/light.ts
import { Vector3 } from '@/lib/math/common';

import { Animation, AnimationOrStatic, animationToString } from '../animation';
import { f, fVector } from '../formatter';
import { Node, nodeAnimations, nodeHeaders } from './node';

export enum LightType {
  Omnidirectional = 'Omnidirectional',
  Directional = 'Directional',
  Ambient = 'Ambient',
}

export interface Light extends Node {
  type: 'Light'
  lightType: LightType;
  attenuationStart: AnimationOrStatic<number>;
  attenuationEnd: AnimationOrStatic<number>;
  intensity: AnimationOrStatic<number>;
  color: AnimationOrStatic<Vector3>;
  ambientIntensity: AnimationOrStatic<number>;
  ambientColor: AnimationOrStatic<Vector3>;
  visibility?: Animation<number>;
}

export function lightsToString(lights: Light[]): string {
  return lights.map((l) => `
    Light "${l.name}" {
      ${nodeHeaders(l)}
      ${l.lightType},

      ${animationOrStaticNumber('AttenuationStart', l.attenuationStart)}
      ${animationOrStaticNumber('AttenuationEnd', l.attenuationEnd)}
      ${animationOrStaticNumber('Intensity', l.intensity)}
      ${animationOrStaticColor('Color', l.color)}
      ${animationOrStaticNumber('AmbIntensity', l.ambientIntensity)}
      ${animationOrStaticColor('AmbColor', l.ambientColor)}
      ${animationToString('Visibility', l.visibility)}

      ${nodeAnimations(l)}
    }`).join('\n');
}

function animationOrStaticNumber(type: string, value: AnimationOrStatic<number>): string {
  return 'static' in value ? `static ${type} ${f(value.value)},` : animationToString(type, value);
}

function animationOrStaticColor(type: string, value: AnimationOrStatic<Vector3>): string {
  return 'static' in value ? `static ${type} { ${fVector(value.value)} },` : animationToString(type, value);
}

File: golang/internal/formats/mdl/components/light.go
package components

import (
	"strings"

	"github.com/pqhuy98/wow-converter/internal/math"
)

type LightType string

const (
	LightOmnidirectional LightType = "Omnidirectional"
	LightDirectional     LightType = "Directional"
	LightAmbient         LightType = "Ambient"
)

type Light struct {
	NodeBase
	LightType         LightType
	AttenuationStart  AnimatedOrStatic[float64]
	AttenuationEnd    AnimatedOrStatic[float64]
	Intensity         AnimatedOrStatic[float64]
	Color             AnimatedOrStatic[math.Vector3]
	AmbientIntensity  AnimatedOrStatic[float64]
	AmbientColor      AnimatedOrStatic[math.Vector3]
	Visibility        *Animation
}

func NewLight(name string) *Light {
	return &Light{NodeBase: NodeBase{Name: name, Type: "Light"}}
}

func LightsToString(lights []*Light) string {
	var blocks []string
	for _, l := range lights {
		var b strings.Builder
		b.WriteString("Light \"")
		b.WriteString(l.Name)
		b.WriteString("\" {\n")
		b.WriteString(NodeHeaders(&l.NodeBase))
		b.WriteString(string(l.LightType))
		b.WriteString(",\n\n")
		b.WriteString(animationOrStaticNumber("AttenuationStart", l.AttenuationStart))
		b.WriteString("\n")
		b.WriteString(animationOrStaticNumber("AttenuationEnd", l.AttenuationEnd))
		b.WriteString("\n")
		b.WriteString(animationOrStaticNumber("Intensity", l.Intensity))
		b.WriteString("\n")
		b.WriteString(animationOrStaticColor("Color", l.Color))
		b.WriteString("\n")
		b.WriteString(animationOrStaticNumber("AmbIntensity", l.AmbientIntensity))
		b.WriteString("\n")
		b.WriteString(animationOrStaticColor("AmbColor", l.AmbientColor))
		b.WriteString("\n")
		b.WriteString(AnimationToString("Visibility", l.Visibility))
		b.WriteString("\n")
		b.WriteString(NodeAnimations(&l.NodeBase))
		b.WriteString("}")
		blocks = append(blocks, b.String())
	}
	return strings.Join(blocks, "\n")
}

func animationOrStaticNumber(typeName string, value AnimatedOrStatic[float64]) string {
	if value.Static {
		return "static " + typeName + " " + FVal(value.Value) + ","
	}
	return AnimationToString(typeName, value.Anim)
}

func animationOrStaticColor(typeName string, value AnimatedOrStatic[math.Vector3]) string {
	if value.Static {
		return "static " + typeName + " { " + FVector3(value.Value) + " },"
	}
	return AnimationToString(typeName, value.Anim)
}
